src/auth/hmacsha256/core/qgsauthhmacsha256method.cpp

```diff
@@ -0,0 +1,172 @@
+/***************************************************************************
+    qgsauthhmacsha256method.cpp
+    --------------------------
+    begin                : January 2022
+    copyright            : (C) 2022 by Vincent Cloarec
+    author               : Vincent Cloarec
+    email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsauthhmacsha256method.h"
+
+#include <QMessageAuthenticationCode>
+
+#include "qgsauthmanager.h"
+#include "qgslogger.h"
+#include "qgsapplication.h"
+
+#ifdef HAVE_GUI
+#include "qgsauthhmacsha256edit.h"
+#endif
+
+
+const QString QgsAuthHmacSha256Method::AUTH_METHOD_KEY = QStringLiteral( "HmacSha256" );
+const QString QgsAuthHmacSha256Method::AUTH_METHOD_DESCRIPTION = QStringLiteral( "HMAC SHA256" );
+const QString QgsAuthHmacSha256Method::AUTH_METHOD_DISPLAY_DESCRIPTION = tr( "HMAC SHA256" );
+
+QMap<QString, QgsAuthMethodConfig> QgsAuthHmacSha256Method::sAuthConfigCache = QMap<QString, QgsAuthMethodConfig>();
+
+
+QgsAuthHmacSha256Method::QgsAuthHmacSha256Method()
+{
+  setVersion( 1 );
+  setExpansions( QgsAuthMethod::NetworkRequest );
+  setDataProviders( QStringList()
+                    << QStringLiteral( "wms" )
+                    << QStringLiteral( "vectortile" ) );
+
+}
+
+QString QgsAuthHmacSha256Method::key() const
+{
+  return AUTH_METHOD_KEY;
+}
+
+QString QgsAuthHmacSha256Method::description() const
+{
+  return AUTH_METHOD_DESCRIPTION;
+}
+
+QString QgsAuthHmacSha256Method::displayDescription() const
+{
+  return AUTH_METHOD_DISPLAY_DESCRIPTION;
+}
+
+bool QgsAuthHmacSha256Method::updateNetworkRequest( QNetworkRequest &request, const QString &authcfg,
+    const QString &dataprovider )
+{
+  Q_UNUSED( dataprovider )
+  const QgsAuthMethodConfig mconfig = getMethodConfig( authcfg );
+  if ( !mconfig.isValid() )
+  {
+    QgsDebugMsg( QStringLiteral( "Update request config FAILED for authcfg: %1: config invalid" ).arg( authcfg ) );
+    return false;
+  }
+
+  const QString key = mconfig.config( QStringLiteral( "key" ) );
+  const QString token = mconfig.config( QStringLiteral( "token" ) );
+
+  QString baseUrl = request.url().toString();
+
+  QString keyedUrl = baseUrl + QStringLiteral( "?key=" ) + key;
```

Comment:
I think we should not convert the URL to a string and then append `?key=` - in theory the URL may already contain some query parameters (e.g. `http://.../{z}/{x}/{y}.png?abc=def`) so in such case there should be `&` instead of `?` added. Best to stick to QUrl and QUrlQuery classes that will handle these parsing bits.